### PR TITLE
Enable quad and octal mode in spi_nor.c to support QPI Flash and Octal Flash

### PR DESCRIPTION
**Pull request by rogeryou**
This PR is to support SPI extended modes, like 1-1-2/1-2-2/1-1-4/1-4-4/1-4D-4D/4-4-4/8-8-8/8D-8D-8D.

1-4D-4D :
 1 means command is sent on 1 line in STR mode；
the first 4D means address is sent on 4 lines in DTR mode；
the second 4D means data is sent on 4 lines in DTD mode；

8D-8D-8D :
the first 8D means command is sent on 8 lines in DTR mode；
the second 8D means address is sent on 8 lines in DTR mode；
the third 8D means data is sent on 8 lines in DTD mode；

### CODE REVIEW

File: dts/bindings/mtd/jedec,spi-nor.yaml [hunk 1]
```diff
@@ -1,26 +1,121 @@
-# Copyright (c) 2018 Peter Bigot Consulting, LLC
-# Copyright (c) 2019 Nordic Semiconductor ASA
+# Copyright (c) 2021, STMicroelectronics
 # SPDX-License-Identifier: Apache-2.0
 
 description: |
-  Properties supporting Zephyr spi-nor flash driver (over the Zephyr SPI
-  API) control of serial flash memories using the standard M25P80-based
-  command set.
+    STM32 OSPI Flash controller supporting the JEDEC CFI interface
```
Comment (erwango) [inline dts/bindings/mtd/jedec,spi-nor.yaml:5]: You're taking a generic binding to make it STM32 specific, that is for sure a "no, never".

File: dts/bindings/mtd/jedec,spi-nor.yaml [hunk 2]
```diff
@@ -1,26 +1,121 @@
-# Copyright (c) 2018 Peter Bigot Consulting, LLC
-# Copyright (c) 2019 Nordic Semiconductor ASA
+# Copyright (c) 2021, STMicroelectronics
```
Comment (erwango) [inline dts/bindings/mtd/jedec,spi-nor.yaml:1]: What ?

**Review by erwango — CHANGES_REQUESTED** [review]
Before going further in the review, I think this PR should provide motivations on the changes and the direction being approved by respective maintainers.
Set as DNM until then.

EDIT: Removing the "platform: STM32" label, as this PR contains modifications on generic components that should be discussed first.

**Comment by rogeryou** [comment]
> Before going further in the review, I think this PR should provide motivations on the changes and the direction being approved by respective maintainers. Set as DNM until then.
> 
> EDIT: Removing the "platform: STM32" label, as this PR contains modifications on generic components that should be discussed first.

OK！Thank you for your review. I will do it.

File: drivers/flash/Kconfig.nor [hunk 3]
```diff
@@ -5,6 +5,7 @@ DT_COMPAT_JEDEC_SPI_NOR := jedec,spi-nor
 
 menuconfig SPI_NOR
 	bool "SPI NOR Flash"
+	default y
```
Comment (nvlsianpu) [inline drivers/flash/Kconfig.nor:8]: this property should depends on existence of DT node which describes SPI-NOR memory

**Review by nvlsianpu — CHANGES_REQUESTED** [review]
can this be divided into a few commits? Especial generic spi_nor commit and others

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Agree with @erwango ; the DT bindings changes are not acceptable

File: dts/bindings/mtd/jedec,spi-nor.yaml (hunk 2 above)
Comment (rogeryou) [reply dts/bindings/mtd/jedec,spi-nor.yaml]: Hi,
I have canceled the modification of this file.

**Comment by rogeryou** [comment]
> Agree with @erwango ; the DT bindings changes are not acceptable

Hi,
I have fixed the issue.

**Comment by mbolivar-nordic** [comment]
>  I have fixed the issue.

You need to rebase your changes into a coherent series of commits.

Please watch https://www.youtube.com/watch?v=mnVNLM7ynMQ for a short overview of what is expected.

**Pull request closed without merging**